pkg/nsx/services/subnet/subnet.go

```diff
@@ -0,0 +1,164 @@
+	}
+
+	wg.Add(1)
+	go subnetService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnet, subnetService.SubnetStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetService, err
+	}
+	return subnetService, nil
+}
+
+func (service *SubnetService) CreateOrUpdateSubnet(obj *v1alpha1.Subnet, orgID, projectID, vpcID string) error {
+	nsxSubnet, err := service.buildSubnet(obj)
+	if err != nil {
+		log.Error(err, "failed to build Subnet")
+		return err
+	}
+	existingSubnet := service.SubnetStore.GetByKey(*nsxSubnet.Id)
+	changed := common.CompareResource(SubnetToComparable(existingSubnet), SubnetToComparable(nsxSubnet))
+	if !changed {
+		log.Info("subnet not changed, skip updating", "subnet.Id", *nsxSubnet.Id)
+		return nil
+	}
+	// TODO Hardcode orgID=default
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, orgID, projectID, vpcID)
+	if err != nil {
+		log.Error(err, "WrapHierarchySubnet failed")
+		return err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return err
+	}
+	// TODO Hardcode ordID=default
+	// Get Subnet from NSX after patch operation as NSX renders several fields like `path`/`parent_path`.
+	if *nsxSubnet, err = service.NSXClient.SubnetsClient.Get(orgID, projectID, vpcID, *nsxSubnet.Id); err != nil {
+		return err
+	}
+	if err = service.SubnetStore.Operate(nsxSubnet); err != nil {
+		return err
+	}
+	log.Info("successfully updated nsxSubnet", "nsxSubnet", nsxSubnet)
+	return nil
+}
+
+func (service *SubnetService) DeleteSubnet(obj interface{}, orgID, projectID, vpcID string) error {
+	var nsxSubnet *model.VpcSubnet
+	switch subnet := obj.(type) {
+	case *v1alpha1.Subnet:
+		var err error
+		nsxSubnet, err = service.buildSubnet(subnet)
+		if err != nil {
+			log.Error(err, "failed to build Subnet")
+			return err
+		}
+	case types.UID:
+		subnets := service.SubnetStore.GetByIndex(common.TagScopeSubnetCRUID, string(subnet))
+		if len(subnets) == 0 {
+			log.Info("subnet is not found in store, skip deleting it", "uid", string(subnet))
+			return nil
+		}
+		nsxSubnet = &subnets[0]
+	}
+	nsxSubnet.MarkedForDelete = &MarkedForDelete
+	// WrapHighLevelSubnet will modify the input subnet, make a copy for the following store update.
+	subnetCopy := *nsxSubnet
+	// TODO Hardcode orgID=default
+	orgRoot, err := service.WrapHierarchySubnet(nsxSubnet, orgID, projectID, vpcID)
+	if err != nil {
+		return err
+	}
+	if err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam); err != nil {
+		return err
+	}
+	if err = service.SubnetStore.Operate(&subnetCopy); err != nil {
+		return err
+	}
+	log.Info("successfully deleted  nsxSubnet", "nsxSubnet", nsxSubnet)
+	return nil
+}
+
+// GetAvailableIPNum TODO: need refactor
+func (service *SubnetService) GetAvailableIPNum(subnet *v1alpha1.Subnet) (int64, error) {
+	if subnet.Spec.DHCPConfig.EnableDHCP {
```

Comment:
there is an API GET https://&lt;policy-mgr&gt;/policy/api/v1/orgs/org-1/projects/project-1/vpcs/vpc-1/subnets/subnet-1/ip-pools/default-ipv4, which can return pool_usage in the response. Could you investigate that if we can use that API to check the subnet usage?

Reply:
ok, I think you mean `/parent_path/subnets/subnet-1/ip-pools/static-ipv4-default`. There is a question that should we enable DHCP for subnet? Subnet's ip-pool is separated into `static-ipv4-default` ip-pool and DHCP ip-pool if DHCP of subnet is enabled.

Reply:
Updated to use static ip-pool `static-ipv4-default` to calculate available IP number as DHCP is not enabled in Subnet.